torch/_inductor/index_propagation.py

```diff
@@ -176,22 +176,32 @@ def unwrap(self, a: Union[Any, IndexPropVar]) -> Any:
 
         return a.value
 
-    def wrap(self, a: Any) -> Union[IndexPropVar, Sequence[IndexPropVar]]:
+    # FIXME MYPYNOFOLLOW thinks this overload clashes with the one below because it doesn't follow the import of
+    # sympy, so it treats sympy.Symbol as an Any type. We can probably remove the type-ignore once we remove the nofollow.
```

Comment:
does importing `Symbol` from `sympy` help with this ? alternatively, is enabling MYPY FOLLOW for this file easier ?

Reply:
> does importing Symbol from sympy help with this ?

Nope, `follow_imports=skip` in mypy-nofollow.ini ensures that no imports are ever followed.

> alternatively, is enabling MYPY FOLLOW for this file easier ?

Lintrunner seems to only support globbing files via include-then-exclude, i.e. the set difference of `include_patterns - exclude_patterns`. The regular MYPY config includes all torch/ files, then excludes all torch/_inductor files. So if I wanted to include just one `_inductor/` file I would have to explicitly exclude all the other inductor files that don't typecheck in follow-imports mode, which is kind of awkward.

I could set up a separate lintrunner section just for this file, but that seems like overkill.

Setting `follow_imports=normal` in mypy-nofollow.ini makes this file typecheck cleanly, but unsurprisingly causes a deluge of errors in the other inductor files. Side note: A lot of those typing errors look legit, i.e. at least mis-typing if not an actual runtime error. We should really drop the 'nofollow' thing as soon as possible...

Reply:
That said, all this `@overload`-ing avoids just one type error in the body of `wrap`. It would be a lot more succinct to suppress that one error via a type-ignore comment instead of specifying these overloads, which in any case require their own type-ignore comment, plus a comment block explaining why. We could always add the overloads back when we enable import following.

I'm actually leaning towards this option now. Wdyt?

Reply:
Scratch all of the above, I've redone the typing to a form that's hopefully better and requires no type-ignore comments at all. I've added a recursive type def `IndexPropExpr` that covers all the possible return values of `wrap`. I'm not sure `IndexPropExpr` is the best name for it though... do you or @peterbell10 have any thoughts?

Reply:
It's not really an expression, maybe `IndexPropResult` would make more sense?